TrackerMC/inc/StrawDigiBundle.hh

```diff
@@ -0,0 +1,35 @@
+// Bundle together references to various StrawDigi products, to circumvent
+// triplicated-loops which pop up when iterating over digis
+//
+// Ed Callaghan, 2023
+
+#ifndef TrackerMC_StrawDigiBundle_hh
+#define TrackerMC_StrawDigiBundle_hh
+
+// mu2e
+#include <Offline/RecoDataProducts/inc/StrawDigi.hh>
+#include <Offline/MCDataProducts/inc/StrawDigiMC.hh>
+
+namespace mu2e{
+  class StrawDigiBundle{
+    public:
+      StrawDigiBundle(const StrawDigi digi, const StrawDigiADCWaveform adcs, const StrawDigiMC mc);
+      StrawDigiBundle(const StrawDigi digi, const StrawDigiADCWaveform adcs);
+      StrawDigiBundle(const StrawDigiBundle& bundle);
+
+      const StrawDigi& GetStrawDigi() const;
+      const StrawDigiADCWaveform& GetStrawDigiADCWaveform() const;
+      const StrawDigiMC& GetStrawDigiMC() const;
+
+      // interface for sorting into buckets of overlapping digitization windows
+      const double time() const;
+    protected:
+      const StrawDigi _digi;
```

Comment:
I am confused why the data members of StrawDigiBundle are const.  It suggests that a deep copy isn't necessary.  Could they be const&?  That would avoid copying data.

The reason I think const& doesn't work is because you allow a null MC later on.  Null objects are generally to be avoided.
For now this is OK, but I would like you to revisit the memory model at some point to see if these copies could be avoided, perhaps by having 2 classes, StrawDigiBundle and StrawDigiBundleMC

Reply:
This is an important issue to resolve soon - but it can be in a future PR.   I will merge now.

For future reference.   There is an idea called the "Facade Pattern", discussed  in the classic "Design Patterns" book, https://www.oreilly.com/library/view/design-patterns-elements/0201633612/ .  Also discussed in the old art wiki: https://cdcvs.fnal.gov/redmine/projects/art/wiki/Data_Product_Design_Guide under "prefer facade classes to transient data members".   This class does more than a simple facade but it does share some ideas.  Have a look and see if any of that advice is useful here.